[senseenergy] Address reconnect issues on failure

jsjames opened the pull request:
Fixes an issue where the senseenergy monitor goes offline due to a network issue - but then does not come back online after the network connection is restored.

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyMonitorHandler.java:
```diff
@@ -204,7 +204,7 @@ public void heartbeat() {
 
         if (thingStatus == ThingStatus.OFFLINE
-                && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATION_ERROR) {
+                && getThing().getStatusInfo().getStatusDetail() != ThingStatusDetail.CONFIGURATION_ERROR) {
```
Does this change matter now that `handleApiException` is changed?
What other statusdetails would the heartbeat see here other then `COMMUNICATION_ERROR` that it should call `goOnline()` ?

jsjames replied:
For the specific issue I was seeing, it would not matter, however, I think this is more robust for unforeseen Exceptions.  My thinking is that if there is any exception where it goes offline (with the exception of a CONFIGURATION_ERROR), it should try to recover.

lsiepel replied:
But all exceptions by `handleApiExceprion` are now set to communication_error.

Without manually checking all code paths, the previous check seems better.

jsjames commented:
Also, I had another structural question.  I ran into a HttpResponseException that was thrown by the request.send() in the sendRequest function (SenseEnergyApi.java).  This is not documented as a possible exception which request.send() can throw.

Is it better/cleaner to wrap all the possible exceptions that the api can throw into the SenseEnergyApiException - which would simplify the code a bit.  Or, should I keep the long list of exceptions that can be thrown for each method declaration?

lsiepel commented:
> Also, I had another structural question. I ran into a HttpResponseException that was thrown by the request.send() in the sendRequest function (SenseEnergyApi.java). This is not documented as a possible exception which request.send() can throw.
> 
> Is it better/cleaner to wrap all the possible exceptions that the api can throw into the SenseEnergyApiException - which would simplify the code a bit. Or, should I keep the long list of exceptions that can be thrown for each method declaration?

Might differ for different cases. But in general I prefer SenseEnergyApiException

In general I usually make three distinctive custom exceptions : config related, transient (timeout) and fatal (connection refused) issues.

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyMonitorHandler.java:
```diff
@@ -361,7 +369,7 @@ private void refreshDevices() {
             senseDevices.entrySet().stream() //
                     .filter(e -> !senseDevicesType.containsKey(e.getKey())) //
                     .forEach(e -> senseDevicesType.put(e.getKey(), deduceDeviceType(e.getValue())));
-        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+        } catch (Exception e) {
```
Why changed this, we prefer the catch of explicit exceptions over the generic `Exception`

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyMonitorHandler.java:
```diff
@@ -230,17 +228,27 @@ && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATI
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR);
+                    break;
+                case DATA:
+                    logger.error("Data exception: {}", e.toString());
```
```suggestion
                    logger.warn("Data exception: {}", e.toString());
```

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyMonitorHandler.java:
```diff
@@ -230,17 +228,27 @@ && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATI
+                    logger.error("Data exception: {}", e.toString());
+                    break;
+                default:
+                    logger.error("SenseEnergyApiException: {}", e.toString());
```
```suggestion
                    logger.warn("SenseEnergyApiException: {}", e.toString());
```

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyMonitorHandler.java:
```diff
@@ -170,7 +168,7 @@ public void goOnline() {
             apiMonitorStatus = getApi().getMonitorStatus(id);
             refreshDevices();
-        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+        } catch (Exception e) {
```
Why changed this, we prefer the catch of explicit exceptions over the generic `Exception`

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyBridgeHandler.java:
```diff
@@ -133,17 +131,27 @@ && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATI
         } else {
-            logger.debug("Unhandled Exception", e);
-            statusDetail = ThingStatusDetail.OFFLINE.NONE;
+            logger.error("Unhandled Exception", e);
```
```suggestion
            logger.warn("Unhandled Exception", e);
```

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyBridgeHandler.java:
```diff
@@ -133,17 +131,27 @@ && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATI
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR);
+                    break;
+                case DATA:
+                    logger.error("Data exception: {}", e.toString());
+                    break;
+                default:
+                    logger.error("SenseEnergyApiException: {}", e.toString());
```
```suggestion
                    logger.warn("Data exception: {}", e.toString());
                    break;
                default:
                    logger.warn("SenseEnergyApiException: {}", e.toString());
```

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyBridgeHandler.java:
```diff
@@ -120,7 +118,7 @@ && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATI
         try {
             getApi().verifyToken();
-        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+        } catch (SenseEnergyApiException e) {
```
Why changed this, we prefer the catch of explicit exceptions over the generic `Exception`

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyBridgeHandler.java:
```diff
@@ -92,7 +90,7 @@ public void initialize() {
         try {
             this.monitorIDs = api.initialize(config.email, config.password);
-        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+        } catch (SenseEnergyApiException e) {
```
Why changed this, we prefer the catch of explicit exceptions over the generic `Exception`

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/api/SenseEnergyApi.java:
```diff
@@ -362,42 +289,55 @@ private Request setHeaders(Request request) {
+        try {
+            logger.trace("REQUEST: {}", request.toString());
+            response = request.send();
+        } catch (Exception e) {
```
`send` can throw multiple types of exceptions, each can lead to a different `SEVERITY`.
Atleast `IOException` and `InteruptException` should be handled different.

lsiepel commented on bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/actions/SenseEnergyMonitorActions.java:
```diff
@@ -105,7 +102,7 @@ Map<String, QuantityType<?>> queryEnergyTrend( //
         try {
             trends = localDeviceHandler.getApi().getTrendData(localDeviceHandler.getId(), trendScale, localDateTime);
-        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+        } catch (Exception e) {
```
Why changed this, we prefer the catch of explicit exceptions over the generic `Exception`

lsiepel reviewed (changes requested):
Thanks, left some comment, specific to the exception handling and logging.

jsjames commented:
@lsiepel - sorry for the delay in getting back to the changes.  I further enhanced the backoff mechanism and hopefully addressed your concerns with the exception handling. Thanks again for your review.

lsiepel commented:
CI build fails on spotless for a comment that needs to be splitted.

jsjames commented:
> CI build fails on spotless for a comment that needs to be splitted.

should be fixed now

lsiepel reviewed (approved):
Thanks. LGTM.

As this touched some vital connection handling; do you need additional time for testing?

jsjames commented:
> Thanks. LGTM.
> 
> As this touched some vital connection handling; do you need additional time for testing?

Its been running for several weeks now - and it looks good.
